### PR TITLE
Side effects and Binds

### PR DESCRIPTION
**Pull request by slushie**
This PR introduces Side Effects on Function Declarations, and exposes a BindClause that consumes those side effects. 

Still a Work in Progress, feedback appreciated!

### CODE REVIEW

File: parser/cst.go [hunk 1]
```diff
@@ -795,17 +806,48 @@ type With struct {
...
-	As    *As    `parser:"@@"`
-	Ident *Ident `parser:"@@"`
-	Func  *FuncDecl
-	Call  *CallStmt
+// BindClause represents the entire "as ..." clause on a CallStmt, with either a
+// Default side effect or a list of Binds.
+type BindClause struct {
+	Pos     lexer.Position
+	As      *As       `parser:"@@"`
+	Default *Ident    `parser:"  @@"`
```
Comment (hinshun) [inline parser/cst.go:814]: Should be `( @@`

File: parser/cst.go [hunk 2]
```diff
@@ -795,17 +806,48 @@ type With struct {
...
+	List    *BindList `parser:"| @@"`
```
Comment (hinshun) [inline parser/cst.go:815]: Should be `| @@ )`

File: parser/cst.go [hunk 3]
```diff
@@ -795,17 +806,48 @@ type With struct {
...
+	default:
+		return shiftPosition(b.Pos, 1, 0)
+	}
+}
+
+// BindList is a parenthetical list of Binds.
+type BindList struct {
+	Pos        lexer.Position
+	OpenParen  *OpenParen  `parser:"@@"`
+	List       []*Bind     `parser:"( ( Newline )? @@ ( \",\" ( Newline )?  @@ )* ( \",\" Newline )? )?"`
```
Comment (hinshun) [inline parser/cst.go:835]: This looks correct.

File: parser/cst.go (hunk 1 above)
Comment (hinshun) [reply parser/cst.go:814]: Also, instead of `Default` I prefer `Ident`

File: parser/cst.go [hunk 4]
```diff
@@ -795,17 +806,48 @@ type With struct {
...
+
+func (b *BindClause) Position() lexer.Position { return b.Pos }
+func (b *BindClause) End() lexer.Position {
+	switch {
+	case b.Default != nil:
+		return b.Default.End()
+	case b.List != nil:
+		return b.List.End()
+	default:
+		return shiftPosition(b.Pos, 1, 0)
```
Comment (hinshun) [inline parser/cst.go:827]: Should be panic?

Comment (slushie) [reply parser/cst.go]: I actually just cargo-culted that line from elsewhere. Changed it to a panic though ;)

File: codegen/chain.go [hunk 5]
```diff
@@ -169,6 +174,15 @@ func (cg *CodeGen) EmitFilesystemBuiltinChainStmt(ctx context.Context, scope *pa
 	}
 
 	switch expr.Name() {
+	case "state":
```
Comment (hinshun) [inline codegen/chain.go:177]: Intentionally left here?

File: checker/builtin.go [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+				obj = &parser.Object{
+					Kind:  parser.DeclKind,
+					Ident: ident,
+					Node: &BuiltinDecl{
+						Ident: ident,
+						Func:  make(map[parser.ObjType]*parser.FuncDecl),
+					},
+				}
+			}
+			node, ok := obj.Node.(*BuiltinDecl)
```
Comment (hinshun) [inline checker/builtin.go:37]: s/node/decl

File: codegen/chain.go [hunk 7]
```diff
@@ -637,12 +651,46 @@ func (cg *CodeGen) EmitFilesystemBuiltinChainStmt(ctx context.Context, scope *pa
...
+				solver.WithCallback(func(resp *client.SolveResponse) error {
+					digest = resp.ExporterResponse[keyContainerImageDigest]
+					return nil
+				}),
+			)
 			if err != nil {
 				return st, err
 			}
+
+			// XXX will this be de-duped if already solved by the binding?
```
Comment (hinshun) [inline codegen/chain.go:665]: Yes, if it's not too complicated we should avoid duping but buildkit should dedupe for us.

File: codegen/codegen.go [hunk 8]
```diff
@@ -305,14 +301,14 @@ func (cg *CodeGen) EmitBlock(ctx context.Context, scope *parser.Scope, typ parse
...
-		chain, err := cg.EmitChainStmt(ctx, scope, typ, call, ac, v)
+		chain, err := cg.EmitChainStmt(ctx, scope, typ, call, v)
 		if err != nil {
 			return v, err
 		}
 
 		var cerr error
 		v, cerr = chain(v)
-		if cerr == nil || cerr == ErrAliasReached {
+		if cerr == nil || errors.As(cerr, &ErrBindingCycle{}) {
```
Comment (hinshun) [inline codegen/codegen.go:311]: Nice, I do like the new `errors.As`.

File: codegen/errors.go [hunk 9]
```diff
@@ -18,6 +18,10 @@ var (
 	// object for a scope. Lexical scope checking should occur in the checker so
 	// that's often where the bug shuold be fixed, but it may be in the codegen.
 	ErrUndefinedReference = errors.Errorf("undefined reference")
+
+	// ErrNoValue is returned when codegen is expecting a side effect value, but
+	// none is available. This is usually caused by bad external dependencies.
+	ErrNoValue = errors.Errorf("no value")
```
Comment (hinshun) [inline codegen/errors.go:24]: This looks unused?

File: solver/solve.go [hunk 10]
```diff
@@ -63,7 +63,7 @@ func WithDownloadOCITarball() SolveOption {
 	}
 }
 
-func WithCallback(fn func() error) SolveOption {
+func WithCallback(fn func(resp *client.SolveResponse) error) SolveOption {
```
Comment (hinshun) [inline solver/solve.go:66]: We should probably make this into a type. Maybe `SolveCallback`? Up to you.

File: codegen/chain.go (hunk 5 above)
Comment (slushie) [reply codegen/chain.go:177]: :facepalm:

**Comment by hinshun** [comment]
@slushie Here's a summary of things I've changed:
- Solve binding immediately instead of waiting till first use (this means that if the binding is unused but the parent function is called, its still solved. if the parent function is not called, then no solves are done)
- Solve binding once for all `EmitBinding`
- Return `ErrBindNoTarget` if you bind like `dockerPush "some/ref" as`
- If decls are duplicated with builtins, show where its duplicated in user code rather than `<builtin>:0:0`
- Fix syntax highlighting in langserver
- Fix some edge case checker/codegen issues and tests

I've tested this against more complicated HLBs and I think this is ready to go! Thank you for this amazing feature! :tada:
